### PR TITLE
compositing: Split Servo up into multiple sandboxed processes.

### PR DESCRIPTION
**Pull request by pcwalton**
Multiprocess mode is enabled with the `-M` switch, and sandboxing is
enabled with the `-S` switch.

r? @metajack

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6884)

### CODE REVIEW

**Comment by pcwalton** [comment]
Note that the processes don't shut down properly yet, for multiple reasons: shutdown channels aren't wired up yet and we don't wait on children, so they become zombies. But I didn't want to make this patch bigger.

**Comment by jdm** [comment]
Review status: 0 of 30 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/devtools/actors/timeline.rs, line 205 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-Jvdm-1VMGc--uGcZblA)** ([raw file](https://github.com/servo/servo/blob/2eb122f394651232abd683fc576a5c4288bf277f/components/devtools/actors/timeline.rs#L205)):</sup>
This change is no longer necessary.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6884)

**Comment by metajack** [comment]
Reviewed 21 of 30 files at r1.
Review status: 21 of 30 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/util/task.rs, line 49 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-JvrMLyMbqVvQwA_Beff)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/util/task.rs#L49)):</sup>
Is there really no way to abstract this over the Sender/IpcSender? This seems like a terrible amount of code duplication.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6884)

**Comment by metajack** [comment]
-S-awaiting-review +S-needs-code-changes 

I think ipc-channel needs more documentation. Some of its uses are pretty hard to grok from the code/comments written.

This looks ok except for the comments I left, which were mostly about docs.

---

Reviewed 9 of 30 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[components/layout/layout_task.rs, line 734 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-JvuI2UhHslTsbyItXez)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/layout/layout_task.rs#L734)):</sup>
This is a quite convoluted way to have a one shot receiver. Please file a followup bug to make a better API for this in ipc-channel. For example, `ROUTER.add_oneshot_route(...)`.

For this patch, please document that a oneshot route is what you're doing here, because it's not clear at all unless you really squint.

---

<sup>**[components/script/dom/htmllinkelement.rs, line 263 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-JvuI_PAGmXj6iVEhGci)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/script/dom/htmllinkelement.rs#L263)):</sup>
Why do we need to clone here? Can't we call `self.script_chan.send(...)` directly?

---

<sup>**[components/script/script_task.rs, line 532 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-JvuNazR1NSQg_L1W5M9)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/script/script_task.rs#L532)):</sup>
Why? We already have a control port. Won't we get these messages without proxying them?

---

<sup>**[components/servo/lib.rs, line 222 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-Jvu9gSyvkOjWwS_nFil)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/servo/lib.rs#L222)):</sup>
This comment is very confusingly written. If "this" is the immediately following block, then the comment is contradictory.

Please find another way to express whatever you are trying to communicate here :)

---

<sup>**[components/servo/lib.rs, line 235 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-JvuAIPl6ori90Ng_caJ)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/servo/lib.rs#L235)):</sup>
Does anything ever force kill this process? Or does this process always hang?

Should we really land this patch with this?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6884)

**Comment by metajack** [comment]
@pcwalton Any updates here? I reviewed this a month ago.

**Comment by pcwalton** [comment]
Rebased up to now. Multiprocess is working but sandboxing is not. This may be Mac OS X Yosemite-specific.

**Comment by pcwalton** [comment]
Review status: 0 of 34 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/htmllinkelement.rs, line 0 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-JvuI_PAGmXj6iVEhGci)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/script/dom/htmllinkelement.rs#L0)):</sup>
Can't do that because that will borrow self. I'll add a comment to make this clearer.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6884)

**Comment by pcwalton** [comment]
About half of the review comments addressed so far.

**Comment by pcwalton** [comment]
Review status: 0 of 33 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/script_task.rs, line 532 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-JvuNazR1NSQg_L1W5M9)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/script/script_task.rs#L532)):</sup>
It's because we select over lots of ports, and we can't select over in-process and IPC ports simultaneously. So we have to proxy IPC ports to in-process ports so that all the ports we're selecting over are of the same type (in this case, in-process ports).


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6884)

**Comment by pcwalton** [comment]
Review status: 0 of 33 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[components/servo/lib.rs, line 222 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6884#-Jvu9gSyvkOjWwS_nFil)** ([raw file](https://github.com/servo/servo/blob/53192458657531e31093157bbd3814998632bd81/components/servo/lib.rs#L222)):</sup>
You're right, this comment makes no sense. I don't even remember what I meant here. Deleting it.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6884)

**Comment by pcwalton** [comment]
Comments addressed. Content processes *should* shut down now, though it's hard to test this since we don't actually evict pipelines from the bfcache!

r? @metajack

**Comment by metajack** [comment]
Much better. r=me when rebased.


---

Reviewed 34 of 34 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6884)

File: components/compositing/Cargo.toml [hunk 1]
```diff
@@ -25,6 +25,9 @@ path = "../script_traits"
 [dependencies.style_traits]
 path = "../style_traits"
 
+[dependencies.style]
+path = "../style"
```
Comment (jdm) [inline components/compositing/Cargo.toml]: This looks like a rebase error to me; there are no new users of the style crate in these changes.

**Comment by metajack** [comment]
@pcwalton Looks like minor update and rebase is all that's needed.

**Comment by pcwalton** [comment]
Rebased. Now depends on #8237

**Comment by jdm** [comment]
#8492 merged! Go go go go go go go!

**Comment by jdm** [comment]
Redone in #8599.

**Pull request closed without merging**